File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        boolean[] visited= new boolean[n]; 
        int count = 0;

        // agar koi node visited NAHI hai, that means count++. kyuki DFS is not able to
        // reach there & it is another province.

        for (int i = 0; i < n; i++) {

                    if (!visited[i]) {
                        count++;
                        dfs(isConnected, i, visited);

                    }                
            }

        return count;

    }

    private void dfs(int[][] isConnected, int node, boolean[] visited) {

        int n = isConnected.length;
        for(int i=0;i<n;i++){
            if(isConnected[node][i]==1 && !visited[i]){
                visited[i]=true; 
                dfs(isConnected, i, visited);

            }
        }

    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution(object):
    def findCircleNum(self, isConnected):
        """
        :type isConnected: List[List[int]]
        :rtype: int
        """
        n = len(isConnected)

        visited = [False]*n
        count=0

        def dfs(node): 
            visited[node] = True 

            for nei in range(n): 
                if visited[nei]==False and isConnected[node][nei]: 
                    dfs(nei)
            
        

        for i in range(n): 
            if not visited[i]: 
                dfs(i)
                count+=1
        return count
    
    # Space = O(n) 
    # Time : O(n2)
            
